### PR TITLE
Extract named constants for magic numbers in BackPropagation.ts

### LINKED ISSUE #2219: Extract named constants for magic numbers in BackPropagation.ts
## Summary

`src/propagate/BackPropagation.ts` contains numerous hardcoded numeric constants that control training behaviour. These magic numbers make it difficult to understand what each value represents and to tune them consistently.

## Specific magic numbers to extract

In `createBackPropagationConfig()` and surrounding functions:
- **Line ~146**: `10` — maximum generations range for random selection
- **Line ~160**: `10_000` — limit bias scale
- **Line ~162**: `100_000` — limit weight scale
- **Lines ~237-240**: `0.95`, `0.2` — learning rate decay factor, coordination factor
- **Lines ~295-305**: `0.95`, `1.0`, `1.3`, `0.5` — adaptive learning rate thresholds and multipliers

## What needs to be done

1. Define named constants at module scope with clear, descriptive names:
   ```typescript
   const MAX_RANDOM_GENERATIONS = 10;
   const DEFAULT_BIAS_LIMIT_SCALE = 10_000;
   const DEFAULT_WEIGHT_LIMIT_SCALE = 100_000;
   const LEARNING_RATE_DECAY_FACTOR = 0.95;
   const COORDINATION_FACTOR = 0.2;
   const ERROR_IMPROVEMENT_THRESHOLD = 0.95;
   const STAGNATION_BOOST_FACTOR = 1.3;
   ```
2. Replace all magic number usages with the named constants
3. Ensure no behavioural change — this is a pure refactor

## Acceptance criteria

- [ ] All magic numbers in BackPropagation.ts are replaced with descriptively named constants
- [ ] No behavioural change (all existing tests pass unchanged)
- [ ] Constants are defined at module scope with clear names
- [ ] `./quality.sh` passes
- [ ] Australian English used throughout

Part of #2215

### PR DESCRIPTION
**Pull request by stservice**
## Summary

Extract named constants for magic numbers in `BackPropagation.ts`. All hardcoded
numeric values controlling training behaviour are now defined as descriptively
named module-scope constants, improving readability and making tuning easier.
This is a pure refactor with no behavioural change. Closes #2219.

## Constants extracted

| Constant                      | Value     | Purpose                                            |
| ----------------------------- | --------- | -------------------------------------------------- |
| `MAX_RANDOM_GENERATIONS`      | `10`      | Upper bound for random generation selection        |
| `DEFAULT_BIAS_LIMIT_SCALE`    | `10_000`  | Default limit scale for bias values                |
| `DEFAULT_WEIGHT_LIMIT_SCALE`  | `100_000` | Default limit scale for weight values              |
| `LEARNING_RATE_DECAY_FACTOR`  | `0.95`    | Default learning rate decay factor                 |
| `COORDINATION_FACTOR`         | `0.2`     | Default bias-weight coordination factor            |
| `ERROR_IMPROVEMENT_THRESHOLD` | `0.95`    | Error ratio below which improvement is significant |
| `ERROR_STAGNATION_THRESHOLD`  | `1.0`     | Error ratio below which stagnation is detected     |
| `IMPROVEMENT_BOOST_FACTOR`    | `1.1`     | Learning rate multiplier on improvement            |
| `STAGNATION_BOOST_FACTOR`     | `1.3`     | Learning rate multiplier on stagnation             |
| `MIN_ERROR_ADJUSTMENT`        | `0.5`     | Minimum adjustment factor when error worsens       |

## Evidence

All 44 existing BackPropagation tests pass unchanged, confirming no behavioural
change.

## Test Plan

- Added `test/propagate/BackPropagationConstants.ts` with 15 tests verifying:
  - Each constant has the correct documented value
  - Default config values use the named constants
  - Random generation stays within `MAX_RANDOM_GENERATIONS` bounds

---

## Automated Fix Details
This PR addresses issue #2219: **Extract named constants for magic numbers in BackPropagation.ts**

## Quality Checks
- Followed TDD methodology
- All new functionality has corresponding tests
- Ran `./quality.sh` - all checks pass

## Notes
- No existing tests were removed or commented out
- If any test modifications were required due to business logic changes, they are documented in the commits

Closes #2219
---

🤖 Processed by: stservice

### CODE REVIEW

**Comment by stservice** [comment]
No code changes required — this is an automated test results report confirming all tests pass.

**Results summary**: 5,506 tests total — 5,468 passed, 38 skipped, 0 failures across 773 suites. All green for commit b5b63cb6.

The PR is in good shape with no test regressions.

**Comment by stservice** [comment]
No code changes required — this is an automated test results report confirming all tests pass.

**Results summary**: 5,506 tests total — 5,468 passed, 38 skipped, 0 failures across 773 suites. All green for commit b5b63cb6.

The PR is in good shape with no test regressions.

**Review by nleck — APPROVED** [review]
